src/storage/index/index_manager.cpp

```diff
@@ -0,0 +1,189 @@
+      key_schema.emplace_back(IndexKeyColumn(col_oid, col_type, col.GetNullable(), INT8_MAX & col.GetAttrSize()));
+    else
+      key_schema.emplace_back(IndexKeyColumn(col_oid, col_type, col.GetNullable()));
+  }
+  index_factory.SetKeySchema(key_schema);
+
+  // Build an empty index
+  return index_factory.Build();
+}
+
+catalog::index_oid_t IndexManager::CreateConcurrently(catalog::db_oid_t db_oid, catalog::namespace_oid_t ns_oid,
+                                                      catalog::table_oid_t table_oid, parser::IndexType index_type,
+                                                      bool unique_index, const std::string &index_name,
+                                                      const std::vector<std::string> &index_attrs,
+                                                      const std::vector<std::string> &key_attrs,
+                                                      transaction::TransactionManager *txn_mgr,
+                                                      catalog::Catalog *catalog) {
+  // First transaction to insert an entry for the index in the catalog
+  transaction::TransactionContext *txn1 = txn_mgr->BeginTransaction();
+  catalog::SqlTableHelper *sql_table_helper = catalog->GetUserTable(txn1, db_oid, ns_oid, table_oid);
+  // user table does not exist
+  if (sql_table_helper == nullptr) {
+    txn_mgr->Abort(txn1);
+    return catalog::index_oid_t(0);
+  }
+  std::shared_ptr<SqlTable> sql_table = sql_table_helper->GetSqlTable();
+  catalog::IndexHandle index_handle = catalog->GetDatabaseHandle().GetIndexHandle(txn1, db_oid);
+
+  // placeholder args
+  catalog::index_oid_t index_oid(catalog->GetNextOid());
+  auto indnatts = static_cast<int32_t>(index_attrs.size());
+  auto indnkeyatts = static_cast<int32_t>(key_attrs.size());
+  bool indisunique = unique_index;
+  bool indisprimary = false;
+  bool indisvalid = false;
+  bool indisready = true;
+  bool indislive = false;
+
+  // Intialize the index
+  Index *index = GetEmptyIndex(txn1, db_oid, index_oid, sql_table.get(), indisunique, key_attrs, catalog);
+  // Initializing the index fails
+  if (index == nullptr) {
+    txn_mgr->Abort(txn1);
+    return catalog::index_oid_t(0);
+  }
+
+  // Add IndexEntry
+  index_handle.AddEntry(txn1, index, index_oid, table_oid, indnatts, indnkeyatts, indisunique, indisprimary, indisvalid,
+                        indisready, indislive);
+
+  // initialize the building flag to false
+  auto index_id = make_index_id(db_oid, ns_oid, index_oid);
+  SetIndexBuildingFlag(index_id, false);
+
+  // Commit first transaction
+  transaction::timestamp_t commit_time = txn_mgr->Commit(txn1, TestCallbacks::EmptyCallback, nullptr);
+
+  // Wait for all transactions older than the timestamp of previous transaction commit
+  // TODO(jiaqizuo): use more efficient way to wait for all previous transactions to complete.
+  while (txn_mgr->OldestTransactionStartTime() < commit_time) {
```

Comment:
For a more efficient solution, I can only come up with using a condition variable that is signaled every time this condition is true. However, I think this will add a lot of complexity, and it requires modification to the transaction manager to know the `commit_time`. I'm not sure if it is worth doing that, or the busy waiting is actually acceptable. 

================

I noticed that `OldestTransactionStartTime` requires grabbing a latch. If so, perhaps using a conditional variable is the way to go.